File: app/src/androidTest/java/com/example/quicknotes/view/SettingsFragmentTest.java
package com.example.quicknotes.view;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.Espresso.pressBack;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;

import android.content.Context;

import androidx.test.ext.junit.rules.ActivityScenarioRule;
import androidx.test.platform.app.InstrumentationRegistry;

import com.example.quicknotes.R;
import com.example.quicknotes.controller.MainActivity;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;

/**
 * Instrumentation test for the SettingsFragment and TagColorSettingsFragment
 * This class tests the functionality of the settings screens,
 * including preference navigation and dialog interactions.
 */
public class SettingsFragmentTest {

    @Rule
    public ActivityScenarioRule<MainActivity> activityRule =
            new ActivityScenarioRule<>(MainActivity.class);

    @Before
    public void setUp() {
        Context ctx = InstrumentationRegistry
                .getInstrumentation()
                .getTargetContext();
        ctx.deleteFile("notes.json");
        navigateToSettings();
    }
    
    /**
     * Helper method to navigate to the settings screen
     */
    private void navigateToSettings() {
        onView(withId(R.id.settingsButton))
                .perform(click());
    }

    
    /**
     * Test that the settings screen displays correctly
     */
    @Test
    public void testSettingsDisplayed() {
        onView(withText("OpenAI API Key"))
                .check(matches(isDisplayed()));

        onView(withText("Auto-Tag Limit"))
                .check(matches(isDisplayed()));

        onView(withText("Allow Notifications"))
                .check(matches(isDisplayed()));

        onView(withText("Delete All Notes"))
                .check(matches(isDisplayed()));

        onView(withText("Edit Tag Colors"))
                .check(matches(isDisplayed()));
    }
    
    /**
     * Test that clicking on the OpenAI API key preference opens a dialog
     */
    @Test
    public void testOpenApiKeyClick() {
        onView(withText("OpenAI API Key"))
                .perform(click());

        onView(withText("OpenAI API Key"))
                .check(matches(isDisplayed()));

        onView(withText("Cancel"))
                .perform(click());
    }
    
    /**
     * Test that clicking on the Delete All Notes preference shows a confirmation dialog
     */
    @Test
    public void testDeleteAllNotesDialog() {
        onView(withText("Delete All Notes"))
                .perform(click());

        onView(withText("Delete All Notes"))
                .check(matches(isDisplayed()));
        
        onView(withText("Are you sure? This will permanently erase all your notes."))
                .check(matches(isDisplayed()));

        onView(withText("Cancel"))
                .perform(click());
    }
    
    /**
     * Test toggling AI mode preference
     */
    @Test
    public void testToggleAiMode() {
        onView(withText("Use AI-powered Auto-Tagging"))
                .perform(click());
        
        onView(withText("Use AI-powered Auto-Tagging"))
                .perform(click());
    }
    
    /**
     * Test navigation to Tag Color Settings screen
     */
    @Test
    public void testNavigateToTagColorSettings() {
        onView(withText("Edit Tag Colors"))
                .perform(click());

        onView(withText("No tags available"))
                .check(matches(isDisplayed()));

        pressBack();
        onView(withText("AI"))
                .check(matches(isDisplayed()));
    }

}

File: app/src/androidTest/java/com/example/quicknotes/view/SettingsFragmentTest.kt
package com.example.quicknotes.view

import androidx.test.espresso.Espresso
import androidx.test.espresso.action.ViewActions
import androidx.test.espresso.assertion.ViewAssertions
import androidx.test.espresso.matcher.ViewMatchers
import androidx.test.ext.junit.rules.ActivityScenarioRule
import androidx.test.platform.app.InstrumentationRegistry
import com.example.quicknotes.R
import com.example.quicknotes.controller.ControllerActivity
import org.junit.Before
import org.junit.Rule
import org.junit.Test

/**
 * Instrumentation test for the SettingsFragment and TagColorSettingsFragment
 * This class tests the functionality of the settings screens,
 * including preference navigation and dialog interactions.
 */
class SettingsFragmentTest {
    @Rule
    var activityRule: ActivityScenarioRule<ControllerActivity?> =
        ActivityScenarioRule<ControllerActivity?>(ControllerActivity::class.java)

    @Before
    fun setUp() {
        val ctx = InstrumentationRegistry
            .getInstrumentation()
            .getTargetContext()
        ctx.deleteFile("notes.json")
        navigateToSettings()
    }

    /**
     * Helper method to navigate to the settings screen
     */
    private fun navigateToSettings() {
        Espresso.onView(ViewMatchers.withId(R.id.settingsButton))
            .perform(ViewActions.click())
    }


    /**
     * Test that the settings screen displays correctly
     */
    @Test
    fun testSettingsDisplayed() {
        Espresso.onView(ViewMatchers.withText("OpenAI API Key"))
            .check(ViewAssertions.matches(ViewMatchers.isDisplayed()))

        Espresso.onView(ViewMatchers.withText("Auto-Tag Limit"))
            .check(ViewAssertions.matches(ViewMatchers.isDisplayed()))

        Espresso.onView(ViewMatchers.withText("Allow Notifications"))
            .check(ViewAssertions.matches(ViewMatchers.isDisplayed()))

        Espresso.onView(ViewMatchers.withText("Delete All Notes"))
            .check(ViewAssertions.matches(ViewMatchers.isDisplayed()))

        Espresso.onView(ViewMatchers.withText("Edit Tag Colors"))
            .check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
    }

    /**
     * Test that clicking on the OpenAI API key preference opens a dialog
     */
    @Test
    fun testOpenApiKeyClick() {
        Espresso.onView(ViewMatchers.withText("OpenAI API Key"))
            .perform(ViewActions.click())

        Espresso.onView(ViewMatchers.withText("OpenAI API Key"))
            .check(ViewAssertions.matches(ViewMatchers.isDisplayed()))

        Espresso.onView(ViewMatchers.withText("Cancel"))
            .perform(ViewActions.click())
    }

    /**
     * Test that clicking on the Delete All Notes preference shows a confirmation dialog
     */
    @Test
    fun testDeleteAllNotesDialog() {
        Espresso.onView(ViewMatchers.withText("Delete All Notes"))
            .perform(ViewActions.click())

        Espresso.onView(ViewMatchers.withText("Delete All Notes"))
            .check(ViewAssertions.matches(ViewMatchers.isDisplayed()))

        Espresso.onView(ViewMatchers.withText("Are you sure? This will permanently erase all your notes."))
            .check(ViewAssertions.matches(ViewMatchers.isDisplayed()))

        Espresso.onView(ViewMatchers.withText("Cancel"))
            .perform(ViewActions.click())
    }

    /**
     * Test toggling AI mode preference
     */
    @Test
    fun testToggleAiMode() {
        Espresso.onView(ViewMatchers.withText("Use AI-powered Auto-Tagging"))
            .perform(ViewActions.click())

        Espresso.onView(ViewMatchers.withText("Use AI-powered Auto-Tagging"))
            .perform(ViewActions.click())
    }

    /**
     * Test navigation to Tag Color Settings screen
     */
    @Test
    fun testNavigateToTagColorSettings() {
        Espresso.onView(ViewMatchers.withText("Edit Tag Colors"))
            .perform(ViewActions.click())

        Espresso.onView(ViewMatchers.withText("No tags available"))
            .check(ViewAssertions.matches(ViewMatchers.isDisplayed()))

        Espresso.pressBack()
        Espresso.onView(ViewMatchers.withText("AI"))
            .check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
    }
}